lib/avromatic/model/types/union_type.rb

```diff
@@ -76,10 +76,16 @@ def referenced_model_classes
         private
 
         def find_index(value)
-          # TODO: Cache this?
-          member_types.find_index do |member_type|
-            member_type.value_classes.any? { |value_class| value.is_a?(value_class) }
+          matched = member_types.find_index do |member_type|
+            member_type.explicitly_coerced?(value)
           end
+
+          # if no explicitly matched found we try to match by the types which we can coerce
+          matched = member_types.find_index do |member_type|
+            member_type.explicitly_coercible?(value)
+          end if matched.nil?
+
+          matched
```

Comment:
It's better to use `coerced?` check directly here:

```ruby
member_types.find_index { |member_type| member_type.coerced?(value) }
```

Seems to be enough to pass your specs.

Reply:
coerced its not always correct - I added matched? helper for that we should cover that.

Reply:
Can you add a test case for the situation when `coerced?` is not correct?

Reply:
-[null, int],  
-[string, map(string)] - in that case there is a spec to support Hash with symbol keys

Reply:
coerced? support nil values which should not be allowed if null is not provided I guess

Reply:
But `serialize` for union_type rules out `nil` way before `find_index` is used, so it shouldn't be relevant.

Reply:
👍 so nil is covered.. so generally this avoids additional nil check.
and what to do with the map(string) when eg map with symbols is passed  ?

Reply:
Let's add a test for that to show the use-case then.

Reply:
actually, there is a test for that that is why I covered that this way

Reply:
🤔 all tests passed when I used `coerced?` here. Anyway, let's keep `matched?`